crates/miden-objects/src/batch/proposed_batch.rs

```diff
@@ -131,27 +133,41 @@ impl ProposedBatch {
         // Verify block header and chain MMR match.
         // --------------------------------------------------------------------------------------------
 
-        if chain_mmr.chain_length() != block_header.block_num() {
+        if chain_mmr.chain_length() != reference_block_header.block_num() {
             return Err(ProposedBatchError::InconsistentChainLength {
-                expected: block_header.block_num(),
+                expected: reference_block_header.block_num(),
                 actual: chain_mmr.chain_length(),
             });
         }
 
         let hashed_peaks = chain_mmr.peaks().hash_peaks();
-        if hashed_peaks != block_header.chain_root() {
+        if hashed_peaks != reference_block_header.chain_root() {
             return Err(ProposedBatchError::InconsistentChainRoot {
-                expected: block_header.chain_root(),
+                expected: reference_block_header.chain_root(),
                 actual: hashed_peaks,
             });
         }
 
         // Verify all block references from the transactions are in the chain MMR, except for the
         // batch's reference block.
+        //
+        // Note that some block X is only added to the blockchain MMR by block X + 1. This is
+        // because block X cannot compute its own block commitment and thus cannot add itself to the
+        // chain. So, more generally, a parent block is added to the blockchain by its child block.
+        //
+        // The reference block of a batch may be the latest block in the chain and, as mentioned,
+        // block is not yet part of the blockchain MMR, so its inclusion cannot be proven. Since the
+        // inclusion cannot be proven in all cases, the batch kernel instead commits to this
+        // reference block's commitment as a public input, which means the block kernel will prove
+        // this block's inclusion when including this batch and verifying its ZK proof.
+        //
+        // Finally, note that we don't verify anything cryptographically here. We have previously
+        // verified that the batch reference block's chain root matches the hashed peaks of the
+        // `ChainMmr`, and so we only have to check if the chain MMR contains the block here.
         // --------------------------------------------------------------------------------------------
```

Comment:
@Mirko-von-Leipzig Let me know if this explanation makes sense.

For your second question from https://github.com/0xPolygonMiden/miden-base/pull/1168#discussion_r1963818058:
> Given that the latest block cannot be proven, does this mean content reference blocks are also skipped if they don't have an inclusion proof? I think you've said as much.

No, they're not skipped. They must be proven by the batch which is done with the chain MMR here, except if the transaction's reference block is the same as the batch reference block, in which case the batch reference block's verification also prove the transaction's ref block.

Reply:
Makes sense, thanks!

> No, they're not skipped. They must be proven by the batch which is done with the chain MMR here, except if the transaction's reference block is the same as the batch reference block, in which case the batch reference block's verification also prove the transaction's ref block.

I meant in the latter case 👍

Reply:
> Finally, note that we don't verify anything cryptographically here. We have previously
> verified that the batch reference block's chain root matches the hashed peaks of the
> `ChainMmr`, and so we only have to check if the chain MMR contains the block here.

Currently we check that chain MMR contains the tx's reference block; but why don't we have to check the hash matches?

Reply:
If the chain MMR's hashed peaks matches the expected chain root, then any block tracked in the chain MMR must be the right one, otherwise the root would be different. Similar to how a partial merkle tree's root would be different from the corresponding full tree's root if we tracked a value that is not actually in the full tree.

Checking whether the chain MMR contains a block with the transaction's block reference number should be equivalent to checking whether the hash of the block is in it, because the chain MMR _should_ be constructed in such a way that the provided headers match the tracked block numbers. At least, that's how I thought it worked, but https://github.com/0xPolygonMiden/miden-base/issues/1173.

Reply:
I assumed as much initially but then I got confused because the hashes aren't anywhere :D